numpyro/infer/reparam.py

```diff
@@ -26,6 +26,93 @@ def __call__(self, name, fn, obs):
         """
         return fn, obs
 
+    def _unwrap(self, fn):
+        """
+        Unwrap Independent(...) distributions.
+        """
+        event_dim = fn.event_dim
+        while isinstance(fn, dist.Independent):
+            fn = fn.base_dist
+        return fn, event_dim
+
+    def _wrap(self, fn, event_dim):
+        """
+        Wrap in Independent distributions.
+        """
+        if fn.event_dim < event_dim:
+            fn = fn.to_event(event_dim - fn.event_dim)
+        assert fn.event_dim == event_dim
+        return fn
+
+    def _unexpand(self, fn):
+        """
+        Unexpand ExpandedDistribution(...) distributions.
+        """
+        batch_shape = fn.batch_shape
+        if isinstance(fn, dist.ExpandedDistribution):
+            fn = fn.base_dist
+        return fn, batch_shape
+
+
+class LocScaleReparam(Reparam):
+    """
+    Generic decentering reparameterizer [1] for latent variables parameterized
+    by ``loc`` and ``scale`` (and possibly additional ``shape_params``).
+
+    This reparameterization works only for latent variables, not likelihoods.
+
+    **References:**
+
+    1. *Automatic Reparameterisation of Probabilistic Programs*,
+       Maria I. Gorinova, Dave Moore, Matthew D. Hoffman (2019)
+
+    :param float centered: optional centered parameter. If None (default) learn
+        a per-site per-element centering parameter in ``[0,1]``. If 0, fully
+        decenter the distribution; if 1, preserve the centered distribution
+        unchanged.
+    :param shape_params: list of additional parameter names to copy unchanged from
+        the centered to decentered distribution.
+    :type shape_params: tuple or list
+    """
+    def __init__(self, centered=None, shape_params=()):
+        assert centered is None or isinstance(centered, (int, float))
+        assert isinstance(shape_params, (tuple, list))
+        assert all(isinstance(name, str) for name in shape_params)
+        if isinstance(centered, (int, float)):
+            assert 0 <= centered and centered <= 1
+        self.centered = centered
+        self.shape_params = shape_params
+
+    def __call__(self, name, fn, obs):
+        assert obs is None, "LocScaleReparam does not support observe statements"
+        centered = self.centered
+        if is_identically_one(centered):
+            return name, fn, obs
+        event_shape = fn.event_shape
+        fn, event_dim = self._unwrap(fn)
+        fn, batch_shape = self._unexpand(fn)
```

Comment:
I am not convinced that the shape of the centered param is compatible with `_unwrap`ping here. I think we may have different understandings of the typical use cases of batch_shape.

Does it make sense to you to initialize `centered` to `jnp.full(batch_shape + event_shape, 0.5)` below?

Reply:
Assume that `fn` is `dist.Normal(0, 1).expand([2]).to_event(1)`. In Pyro, it is the same as `dist.Normal(jnp.zeros(2), jnp.ones(2)).to_event(1)`. After `unwrap`, we get `dist.Normal(jnp.zeros(2), jnp.ones(2))` and `event_dim=1`. Later, we define `centered = jnp.full(event_shape, 0.5)`.

In NumPyro, after `unwrap`, we get `dist.ExpandedDistribution(dist.Normal(0, 1), (2,))` and `event_dim=1`. After `unexpand`, we get `fn=dist.Normal(0, 1)`, `event_dim=1`, `batch_shape=(2,)`. Because `centered` in Pyro is defined independently of `batch_shape`, I think two implementations give the same behavior. WDYT?

Reply:
hmm, I think we need to change `centered` to `full(shape, 0.5)` with `shape = (original fn).shape()` in both Pyro and NumPyro.

Reply:
hmm, it has been a while since I looked at the `LocScaleReparam` interface, but I believe I was thinking:

> Users may want to share statistical strength across various batch dimensions, but probably not across event dimensions. Probably the most common usage pattern for learnable `centered` is to create one learnable parameter per event element, i.e. `pyro.param(..., full(event_shape, 0.5))`. In all less common cases, users can create a `centered = pyro.param(...)` statement outside of the `pyro.reparam()` context and pass the `centered` parameter to the `LocScaleReparam`.

Given that perspective, I am happy with the Pyro interface. Is it possible in NumPyro to add support for tensors, but maybe omit some of the short-circuit optimizations in the Pyro version?

Reply:
Got it! I believe all I need to do is to change the first assertation in the constructor to `isinstance(centered, (ndarray, ndgeneric, DeviceArray, Tracer,...)` and only do validation for `int, float` cases. The `_unexpand` logic here seems not relevant.